collector/processor/concurrentbatchprocessor/batch_processor.go

```diff
@@ -429,8 +432,15 @@ func (b *shard) sendItems(trigger trigger) {
 		// terminates.
 		parentSpan.End()
 
-		for _, pending := range thisBatch {
-			pending.waiter <- countedError{err: err, count: pending.count}
+		if !b.processor.earlyReturn {
+			for _, pending := range thisBatch {
+				select {
+				case pending.waiter <- countedError{err: err, count: pending.count}:
+					// OK! Caller received error and count.
+				case <-pending.ctx.Done():
+					// OK! Caller context was canceled.
```

Comment:
@lquerel This section of code will stop sending countedErrors back to the waiter in case the waiter's context is canceled. There is no more a need to defer an asynchronous function to await for countedErrors after the waiter returns early, because the (a) the countedErrors are only needed while still waiting, (b) the context is available here to know when the waiter stops waiting.